### ISSUE TITLE
(Question) - Offload functions dynamically

### ISSUE THREAD

**Issue by laterbreh** [josdejong/workerpool#17]
Hi, I am not sure if this is the right place to ask this question, but its regarding "Offload functions dynamically". 

Does this mean that a worker will be created every time that is called (following Offload functions dynamically example)? So lets say I have 1 person initiate a task that gets called according to Offload functions dynamically example, and while that worker is processing, another person initiates the same task. 

Will it spawn another worker while the other is still processing? So if two people say initiate the task, it would spawn two workers to process that data for those two requests? (I'm using this in an express application).

Thank you in advance!

**Comment by josdejong** [comment]
Thanks and good question, thanks for asking. 

In the [offloadFunctions](https://github.com/josdejong/workerpool/blob/master/examples/offloadFunctions.js) example, every time you call `pool.exec(add, [3, 4])`, the function `add` will be serialized and send to the worker to execute, together with the arguments `[3, 4]`. After the function is executed, the result is returned and the worker completely forgets about the function `add`.

This is completely separate from the decision of the workerpool whether or not to spawn a new worker or not, the flow goes like this:
- A new task comes in (doesn't matter whether a dedicated task or offloaded function)
- workerpool checks whether there is an idle worker available to do the task, if so the task is passed to this worker
- when there is no idle worker, the workerpool checks whether the number of current workers is below `maxWorkers` and if so, it spawns a new worker and passes the task to this new worker.
- when the maximum number of workers is instantiated and each of them is busy, the workerpool will put the new task in a queue
- when a worker finishes a task, it checks whether there is a new task available in the queue and if so, it grabs the first/oldest task from the queue and start working on that.

**Comment by laterbreh** [comment]
Thank you for the clarification. 

So if the example structure for dynamic offloading is followed, the default amount of available workers is NumCPUs - 1, by default if none is specified. 

My last question would be then, Would you recommend this library for use in an express application? I thought to use this to move all my blocking functions outside of the servers main event loop. Does this sound like a proper use-case? I ask this because if the number of workers dynamically set according to the amount of available CPUs, under heavy load, the CPU bound activities are still bound by the number of CPUs (in dynamic offloading with no maximum set). 

Thank you again for your responses!

**Comment by josdejong** [comment]
> So if the example structure for dynamic offloading is followed, the default amount of available workers is NumCPUs - 1, by default if none is specified.

That's correct

> My last question would be then, Would you recommend this library for use in an express application?

Yes, that's exactly the sort of use cases I designed `workerpool` for. Move heavy, blocking operations away from the main thread into a worker or child_process (both server side and in the browser). 

You can create more than CPUs - 1 workers if you want, you can create hundreds if you want. They will all run in parallel and you leave it up to your underlying OS and hardware to schedule all these threads over a limited amount of physical CPU's. Having just one process per CPU is normally most efficient because the CPU doesn't have to do context switching all the time. Another concern is memory usage: every worker requires a serious amount of memory, so creating hundreds may not be a good idea.

**Comment by laterbreh** [comment]
Thank you so much for the fantastic clarification. We will be playing with the worker settings as the application grows and we are able to seriously benchmark it. 

Thank you for the clarification. 

On a side note: You may want to drop in these explantations in the README.

**Comment by josdejong** [comment]
You're welcome. Extending the docs can indeed be helpful.
